net/respondd/src/respondd.c

```diff
@@ -320,27 +384,99 @@ static struct json_object * handle_request(char *request, bool *compress) {
 	}
 }
 
-
-static void serve(int sock) {
-	char input[256];
-	const char *output = NULL;
-	ssize_t input_bytes, output_bytes;
+// Wait for an incomming request and schedule it.
+//
+// 1a. If the schedule is empty, we wait infinite time.
+// 1b. If we have scheduled requests, we only wait for incomming requests
+//     until we reach the scheduling deadline.
+// 2a. If the incomming request was sent to a multicast destination IPv6,
+//     choose a random delay between 0 and max_multicast_delay milliseconds.
+// 2b. If the incomming request was sent to a unicast destination, delay
+//     is zero.
+static void accept_request(struct request_schedule *schedule, int sock,
+                           uint64_t max_multicast_delay) {
+	char input[REQUEST_MAXLEN];
+	ssize_t input_bytes;
 	struct sockaddr_in6 addr;
 	socklen_t addrlen = sizeof(addr);
-	bool compress;
+	char control[256];
+	struct in6_addr destaddr;
+	struct cmsghdr *cmsg;
+
+	int64_t timeout = schedule_idle_time(schedule);
+	if (timeout < 0)
+		return;
+
+	// set timeout to the socket
+	struct timeval t;
+	t.tv_sec = ((uint64_t) timeout) / 1000;
+	t.tv_usec = (((uint64_t) timeout) % 1000) * 1000;
+
+	if (setsockopt (sock, SOL_SOCKET, SO_RCVTIMEO, (char *)&t, sizeof(t)) < 0)
+		perror("setsockopt failed\n");
+
+	struct iovec iv = {
+		.iov_base = input,
+		.iov_len = sizeof(input) - 1
+	};
+
+	struct msghdr mh = {
+		.msg_name = &addr,
+		.msg_namelen = sizeof(addr),
+		.msg_iov = &iv,
+		.msg_iovlen = 1,
+		.msg_control = control,
+		.msg_controllen = sizeof(control)
+	};
 
-	input_bytes = recvfrom(sock, input, sizeof(input)-1, 0, (struct sockaddr *)&addr, &addrlen);
+	input_bytes = recvmsg(sock, &mh, 0);
+
+	// Timeout
+	if (input_bytes < 0 && errno == EWOULDBLOCK)
+		return;
 
 	if (input_bytes < 0) {
-		perror("recvfrom failed");
+		perror("recvmsg failed");
 		exit(EXIT_FAILURE);
 	}
 
+	// determine destination address
+	for (cmsg = CMSG_FIRSTHDR(&mh);	cmsg != NULL; cmsg = CMSG_NXTHDR(&mh, cmsg))
+	{
+		// skip other packet headers
+		if (cmsg->cmsg_level != IPPROTO_IPV6 || cmsg->cmsg_type != IPV6_PKTINFO)
+			continue;
+
+		struct in6_pktinfo *pi = (struct in6_pktinfo *) CMSG_DATA(cmsg);
+		destaddr = pi->ipi6_addr;
+		break;
+	}
+
 	input[input_bytes] = 0;
 
-	struct json_object *result = handle_request(input, &compress);
-	if (!result)
-		return;
+	// only multicast requests are delayed
+	update_time();
+	int64_t delay;
+	if (IN6_IS_ADDR_MULTICAST(&destaddr)) {
+		delay = rand() % max_multicast_delay;
+	} else {
+		delay = 0;
+	}
+
+	struct request_task *new_task = malloc(sizeof(struct request_task));
+	new_task->scheduled_time = now + delay;
+	strcpy(new_task->request, input);
+	memcpy(&new_task->client_addr, &addr, addrlen);
+
+	if(!schedule_push_request(schedule, new_task)) {
+		free(new_task);
```

Comment:
I don't like the fact that a full schedule can make respondd drop requests - and it is even possible to break unicast requests my sending to many multicasts!

Let's rather:
* Not queue requests with delay == 0 at all, but just serve them directly
* Serve requests instead of dropping them when the queueing fails